### PR TITLE
Configure scope-based folding for most syntaxes

### PR DESCRIPTION
**Pull request by BenjaminSchaaf**
This supplants `Default/Fold.tmPreferences` entirely and thus the
default will revert to indentation-based folding.

Syntaxes left out of this were lacking the `meta.` scopes or needed
other work to properly configure scope-based folding.

Fixes https://github.com/sublimehq/sublime_text/issues/5348
Fixes https://github.com/sublimehq/sublime_text/issues/5332

### CODE REVIEW

File: C#/Fold.tmPreferences [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                <string>punctuation.section.sequence.begin</string>
+                <key>end</key>
+                <string>punctuation.section.sequence.end</string>
+            </dict>
+            <dict>
+                <key>begin</key>
+                <string>punctuation.section.group.begin</string>
+                <!-- record parameters start as a group -->
+                <key>end</key>
+                <string>punctuation.section.group.end | punctuation.section.parameters.end</string>
```
Comment (deathaxe) [inline C#/Fold.tmPreferences:47]: The mentioned issue was fixed by #3320.

That said, I have a feeling those generous selectors could cause unexpected behavior in other situations. So we should rather try to fix such things in syntax definitions.

```suggestion
                <key>end</key>
                <string>punctuation.section.group.end</string>
```

File: JavaScript/Fold.tmPreferences [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                <string>punctuation.section.block.begin</string>
+                <key>end</key>
+                <string>punctuation.section.block.end</string>
+            </dict>
+            <dict>
+                <key>begin</key>
+                <string>punctuation.section.brackets.begin</string>
+                <key>end</key>
+                <string>punctuation.section.brackets.end</string>
+            </dict>
```
Comment (deathaxe) [inline JavaScript/Fold.tmPreferences:23]: Dupicate entry can be removed.

```suggestion
```

File: Java/Fold.tmPreferences [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    <key>settings</key>
+    <dict>
+        <key>foldScopes</key>
+        <array>
+            <dict>
+                <key>begin</key>
+                <string>punctuation.section.block.begin</string>
+                <key>end</key>
+                <string>punctuation.section.block.end</string>
+            </dict>
```
Comment (deathaxe) [inline Java/Fold.tmPreferences:15]: Folding groups should be safe as well in Java.

```suggestion
            <dict>
                <key>begin</key>
                <string>punctuation.section.block.begin</string>
                <key>end</key>
                <string>punctuation.section.block.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.group.begin</string>
                <key>end</key>
                <string>punctuation.section.group.end</string>
            </dict>            
```

File: JSON/Fold.tmPreferences [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                <string>punctuation.section.sequence.begin</string>
+                <key>end</key>
+                <string>punctuation.section.sequence.end</string>
+            </dict>
+            <dict>
+                <key>begin</key>
+                <string>punctuation.section.mapping.begin</string>
+                <key>end</key>
+                <string>punctuation.section.mapping.end</string>
+            </dict>
```
Comment (deathaxe) [inline JSON/Fold.tmPreferences:23]: Maybe of lower priority, but do we want to find a common scheme of sorting those rules?
```suggestion
            <dict>
                <key>begin</key>
                <string>punctuation.section.mapping.begin</string>
                <key>end</key>
                <string>punctuation.section.mapping.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.sequence.begin</string>
                <key>end</key>
                <string>punctuation.section.sequence.end</string>
            </dict>
```

File: Python/Fold.tmPreferences [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                <string>punctuation.section.set.begin</string>
+                <key>end</key>
+                <string>punctuation.section.set.end</string>
+            </dict>
+            <dict>
+                <key>begin</key>
+                <string>punctuation.section.brackets.begin</string>
+                <key>end</key>
+                <string>punctuation.section.brackets.end</string>
+            </dict>
```
Comment (deathaxe) [inline Python/Fold.tmPreferences:53]: Maybe sort rules, starting with semantic punctuation (arguments, parameters, group, mapping, ...), followed by syntactical (brackets) with each group sorted alphabetically?

```suggestion
            <dict>
                <key>begin</key>
                <string>punctuation.section.group.begin</string>
                <key>end</key>
                <string>punctuation.section.group.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.mapping.begin</string>
                <key>end</key>
                <string>punctuation.section.mapping.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.sequence.begin</string>
                <key>end</key>
                <string>punctuation.section.sequence.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.set.begin</string>
                <key>end</key>
                <string>punctuation.section.set.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.brackets.begin</string>
                <key>end</key>
                <string>punctuation.section.brackets.end</string>
            </dict>
```

Comment (deathaxe) [inline Python/Fold.tmPreferences:53]: Maybe sort rules, starting with semantic punctuation (arguments, parameters, group, mapping, ...), followed by syntactical (brackets) with each group sorted alphabetically?

```suggestion
            <dict>
                <key>begin</key>
                <string>punctuation.section.group.begin</string>
                <key>end</key>
                <string>punctuation.section.group.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.mapping.begin</string>
                <key>end</key>
                <string>punctuation.section.mapping.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.sequence.begin</string>
                <key>end</key>
                <string>punctuation.section.sequence.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.set.begin</string>
                <key>end</key>
                <string>punctuation.section.set.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.brackets.begin</string>
                <key>end</key>
                <string>punctuation.section.brackets.end</string>
            </dict>
```

File: Rust/Fold.tmPreferences [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+                <string>punctuation.section.group.begin</string>
+                <key>end</key>
+                <string>punctuation.section.group.end</string>
+            </dict>
+            <dict>
+                <key>begin</key>
+                <string>punctuation.section.block.begin</string>
+                <key>end</key>
+                <string>punctuation.section.block.end</string>
+            </dict>
```
Comment (deathaxe) [inline Rust/Fold.tmPreferences:23]: Reorder blocks, groups ... ?

```suggestion
            <dict>
                <key>begin</key>
                <string>punctuation.section.block.begin</string>
                <key>end</key>
                <string>punctuation.section.block.end</string>
            </dict>
            <dict>
                <key>begin</key>
                <string>punctuation.section.group.begin</string>
                <key>end</key>
                <string>punctuation.section.group.end</string>
            </dict>
```

**Review by deathaxe — COMMENTED** [review]
With a look at the differences handing folding rules per syntax is definitly the better strategy.

**Comment by keith-hall** [comment]
After these changes, how will users be able to disable scope-based folding and restore indentation-based folding (globally i.e. for all syntaxes)? Presumably https://discord.com/channels/280102180189634562/280102180189634562/978281955806285824 wouldn't work any more because of selector specificity.

**Comment by BenjaminSchaaf** [comment]
> After these changes, how will users be able to disable scope-based folding and restore indentation-based folding (globally i.e. for all syntaxes)?

They won't, the point is to conservatively apply scope-based folding where it always works and improves the status-quo and thus doesn't need an option to be disabled.

**Comment by deathaxe** [comment]
I just wonder how long it takes for issues to be filed about "I can fold parens in syntax A but why doesn't it work in syntax B?"

File: Java/Fold.tmPreferences (hunk 3 above)
Comment (BenjaminSchaaf) [reply Java/Fold.tmPreferences:15]: Unfortunately not, see for example syntax_test_java.java:6137:

```
    for (; {}
//  ^^^^^^^^^ - meta.statement.loop.for meta.statement.loop.for
//  ^^^^ meta.statement.loop.for.java - meta.group
//      ^^^ meta.statement.loop.for.java meta.group.java
//         ^^ meta.statement.loop.for.java meta.block.java
//  ^^^ keyword.control.loop.for.java
//     ^ - keyword
//      ^ punctuation.section.group.begin.java
//       ^ punctuation.terminator.java
//         ^ punctuation.section.block.begin.java
//          ^ punctuation.section.block.end.java
```

File: JSON/Fold.tmPreferences (hunk 4 above)
Comment (BenjaminSchaaf) [reply JSON/Fold.tmPreferences:23]: I don't think there's ever going to be enough for it to really matter.

**Comment by BenjaminSchaaf** [comment]
I'm hoping those will be fairly minor, as most of the time indentation does a fairly good job already. I've addressed all review comments.

**Review by deathaxe — APPROVED** [review]

**Pull request merged**